### PR TITLE
Update to Bazel 4.0.0

### PR DESCRIPTION
**Pull request by aherrmann**
- Update nixpkgs to latest nixpkgs-unstable.
- Use Bazel 4.0.0 for nixpkgs and bindist.
- Bump the version bounds in `start` and `versions.bzl`
- rules_haskell dependencies were just updated recently on https://github.com/tweag/rules_haskell/pull/1487

### CODE REVIEW

**Comment by aherrmann** [comment]
Blocked on https://github.com/bazelbuild/bazel/pull/13044 --> converting to draft.

**Comment by aherrmann** [comment]
I've added a workaround for https://github.com/bazelbuild/bazel/pull/13044 to unblock this: Instances of arguments starting with `@` are transformed into arguments of the form `-Wl,...` in `rules_haskell`'s `cc_wrapper`. E.g. `-Xlinker -rpath -Xlinker @loader_path/... --> -Wl,-rpath,@loader_path/...` and `-library_name @rpath/... --> -Wl,-library_name,@rpath/...`. This is only an issue on MacOS, so the Windows `cc_wrapper` does not need to be modified.

**Review by regnat — APPROVED** [review]
It's makes my idealistic self a bit sad to see that the bazel version has to be specified in so many places. But apart from that, all good for me :+1:

**Pull request merged**
